Bluetooth: Controller: Integrate direction finding connected mode with refactored LLCPs

ppryga-nordic opened the pull request:
The PR is a set of small commits that add missing pieces of code responsible for integration of direction finding connected mode with refactored Link Layer Control Procedures.

The commits touches host, upper link layer, refactored LLCPs and lower link layer.

ppryga-nordic committed:
Bluetooth: Controller: ULL: Add impl for HCI connection CTE rsp enable

There were missing code responsible for execution of HCI_LE_Connection_-
CTE_Response_Enable HCI command.

The commit adds missing implementation into hci and upper link layer.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: ull: df: use new LLCP structus in conn CTE req

The implementation of the ll_df_set_conn_cte_req_enable function was
based on former implementation of LLCPs. The CTE request and response
control procedures are not implemented in former LLCPs framework.
The code has been updated to use data structues from refactored
implementation of LLCPs.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: Add Kconfig to enable CTE RX in conn mode

The CTE reception and sampling in connected mode was enabled
with CONFIG_BT_CTLR_DF_CONN_CTE_REQ. This is a separate fearure
tha can be enabled when CTE reqest procedure is disabled.
What more the CONFIG_BT_CTLR_DF_CONN_CTE_REQ is dependent
on CONFIG_BT_CTLR_DF_CONN_CTE_RX, not other way around.

The commit adds separate Kconfig to provide such possibility.
Also changes compilation guards for code related with the
CTE reception and sampling.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
BLuetooth: Controller: llcp: Fix wrong conditions in remote CTE req

There were wrong conditions in rp_comm_tx function for CTE request
that vefiry if:
- PHY is allowed one,
- CTE length is within allowed value.

The commit fixes it.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: llcp: setting of CTE related content in pdu_data

The llcp_pdu_encode_cte_rsp functio didn't set pdu_data fields related
with CTE transmission. The commit fixes that.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: df: Use PDUs CTEInfo to config radio TX

CTE type and length parameters in connected mode are dependent
on remote CTE request procedure. These parameters are required
to setup radio to correctly transmit CTE. The same parameters
are also included in CTEInfo byte of data channel PDU header.
The parameters were provided to LLL with PDU to be transmitted
and by lll_conn::df_tx_cfg, hence it was redundant.
The PDUs CTEInfo has to be set before transmission anyway.
The contents of the PDU are set in ULL by CTE RSP control
procedure. To remove redundancy CTE length and type from lll_df_-
conn_tx_cfg were removed from ll_df_conn_tx_cfg. It was better
option becuse it saves instructions in LLL.

Radio in connected mode was configured by lll_df_conn_cte_tx_enable.
The function just unpacked the lll_df_conn_tx_cfg mebmers and
called static function df_cte_tx_configure.
Instead of extending parameters list of lll_df_conn_cte_tx_enable,
the function was removed and df_cte_tx_configure was changed to
be global function - lll_df_cte_tx_configure.
Now LLL directly passes all parameters from pdu_data::cte_info
and ll_df_conn_tx_cfg to lll_df_cte_tx_configure.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: hci: Fix wrong CTE request interval type

Request interval is a number of connection events that
is used to periodically run CTE request control procedure.
BT 5.3 Core Specification defines it as 2 octets long.
It had wrong type uint8_t. Changed to correct one uint16_t.

The commit also changes type of cte_rsp_en field of lll_df_conn_tx_cfg
to state that it is a boolean flag.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: Add init CTE REQ and RSP variables on conn create

There were missing initialization for CTX request and response conotrol
procedure related variables. The variables were zeored on system
startup but not initialized when a connection is established.
In case of re-use of connection instance for new connection it was not
possible to setup procedures again.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: llcp: Make remote CTE REQ to compile with CTE RSP

Part of the CTE request procdure that is related with handling of
remote request should be compiled when CTE RSP is enabled.
Withouth it CTE response will not work. Local CTE request does not
need code reponsible for handling of remote reques.

The commit changes conditional compilation guards.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: Remove unused members from llcp_df_rsp_cfg

ant_sw_len and ant_ids members of the structure llcp_df_rsp_cfg
were not used. These data are stored in lll_df_conn_tx_cfg
that is member of lll_conn. Unused members are removed.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: Add storage for remote CTE request data

There are data received from peer device with CTE request.
These data are not part of local CTE request procedure and
they don't belong to CTE response configuration, hence separate
storage was provided.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: Integrate CTE req in ULL with refactored LLCPs

Integrate existing CTE request control procedure code in ULL with
implementation of refactor LLCPs.
The commit includes code responsible for:
- enabling CTE request and scheduling its execution by refactored
  LLCPs framework,
- running CTE request periodically
- disabling CTE reqest in case it is running periodic

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Add handling of CTE request failed

CTE request control procedure may failed due to rejection by
peer device or due to receive of LL_CTE_RSP PDU without CTE.
These events has to be reported to host by HCI_LE_CTE_Request_Failed.

The commit adds missing functionalit.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: host: df: Fix to wrong variable passed to net_buf_add

Wrong variable was passed to net_buff_add call. In could lead to
memory overwrite.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: host: df: Fix wrong arguments order

Function valid_cte_req_params was called with cte_length and
cte_type in wrong order.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: host: df: Add handling of HCI_LE_CTE_Request_Failed

There were no handling of HCI_LE_CTE_Request_Failed event.
The commit adds missing implementation. An application will
be notified about failed request by cte_report_cb. It is the
same callback that is used for reporting collected CTE IQ
samples. The same callback was used to avoid creation new callback.
To give an application possibility to distinguish between regular
IQ samples report and request failed additional member err was added
to bf_df_conn_iq_samples_report structure.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: host: df: Add missing bt_conn_unref

There were missing bt_conn_unref after reporting IQ samples to
an application.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: radio: Fix error in pkt conf macros

Macros responsible for preparation of packet configuration flags in
regard of PHY and CTE were wrong. PHY flags are not used as regular
integer values but bits in a bitfield, hence size of the field in
packet configuration flags is three instead of two.
In such case CTE presen filed should be moved to bit 4th.

The problem was spot when testing implementation with CODED PHY
enabled. When device was sending PDUs that had attached CTE the
radio was configured to use CODED PHY due to wrong bit set in
packet configuration flags variable (overlapping of CTE bit
with CODED PHY bit).

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: llcp: Fix PHY update FSM stale after instant

In peripheral role when:
 - data length update is enabled
 - the CONFIG_BT_MAX_CONN is set to 1
the PHY update control procedure after reaching instant sends two
notifications to host. It notifies host about PHY update and data
length change. Both notifications are send one after another, so two
free node rx are required. The number of available node rx in
provided conditions is one. The PHY update FSM is stalled in waiting
state for enough empty node rx. At the same time remote device is
allowed to send new remote control procedures. Received request are
handled by PHY update FSM that asserts due to unknown procedure opcode.

The commit changes number of allocated nodes to be two times a number
of LLCP connections.

This is a workaround for the issue. The best solution seems to be
sharing nodsx between connections. Though this solution requires
buffering remote control procedures until host notification is done.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: Add packet TX restrictions for CTE REQ/RSP PDUs

There are packet restrictions imposed by PHY update procedure
for PDU that includes CTE (BT Core 5.3 :
- central/peripheral can't send PDU with CTE after receive or send
LLPHY_UPDATE_IND until instant if the PHY after instant is CODED
- peripheral can't send PDU including CTE after it sends LL_PHY_REQ
PDU until receive LL_PHY_UPDATE_IND, LL_UNKNOWN_RSP, LL_REJECTED_EXT_-
IND_PDU if there is a CODE PHY in TX_PHYS
- peripheral can't send PDU including CTE after it sends LL_PHY_RSP
PDU until receive LL_PHY_UPDATE_IND if there is a CODED PHY in TX_PHYS
of LL_PHY_RSP PDU or RX_PHYS of LL_PHY_REQ PDU.

The BT 5.3 Core spec defines only one PDU that may include CTE, that is
LL_CTE_RSP PDU. To avoid a situation that there is such PDU enqueued
for transmission in LLL when packet transmission restrictions should
be applied, both procedures in almost all cases will not be executed
in parallel.

Current implementation always handles remote procedurerequest first.

There are possible three scenarios:
1. Remotely requested PHY update. Locally initiated CTE REQ.
   In this case there is no problem with LL_CTE_RSP waiting in a TX
   queue in LLL. Both procedures may be executed one after another.
2. Remotely requested CTE REQ. Locally initiated PHY update.
   In this case the CTE REQ is handled first and it will pause the
   PHY update procedure until LL_CTE_RSP PDU is acknowledged by
   remote. Then the CTE REQ procedure will be completed and PHY update
   continued.
3. Locally initiated PHY update is pending. Arrives remote CTE REQ.
   In this case the CTE REQ will be paused until localy initiated PHY
   update is completed. Then the CTE REQ will be continued.

Thanks to that there should be no PDU including CTE in LLL TX quueue.
That releases us from a situation there is a LL_CTE_RSP PDU in the
LLL TX qeueue that must be changed into LL_REJECT_EXT_IND PDU due to
change of PHY to CODED PHY after PHY update procedure completes.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: Add missing LLCP collision resolve

When there is pending local control procedure that has instant
it should be possible to run remote control procedure without
instant because there is no collision.

There were missing code to run this case in rr_st_idle() of
refactored LLCPs.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: Bluetooth: Fix failing tests related with direction finding

Fixes build issues for tests in related with CTE REQ and
hci commands.
Fixes issues in unit tests for CTE REQ/RSP control proedures.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
Fixed compliance and tests issues.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -1041,13 +1043,13 @@ uint8_t ll_df_set_conn_cte_tx_params(uint16_t handle, uint8_t cte_types, uint8_t
 
 	df_tx_cfg->cte_types_allowed = cte_types;
-	df_tx_cfg->state = DF_CTE_CONN_TX_PARAMS_SET;
+	df_tx_cfg->is_initialized = true;
```
Is it always typesafe (regardless of compiler) to assign `true` and `false` to a `uint8_t:1` field?

ppryga replied:
It is if a field is bool:1. I've changed the type to `bool` to state that its purpose it to be a flag.

Thalley replied:
But didn't we have that discussion in your earlier PR? 

e.g. if we have
```
bool:1 b_val;
uint8_t:7 u_val;
```
Is that always 1 octet, or will it be 2 octets, and is it the same for all compilers? 

Personally I think that, to be type safe, we should assign `1U` and `0U` if it is `uintX_t:1`, and `true` and `false` if it is `bool:1`, but `bool:1` should perhaps only be used if it can be used together with `uintX_t` bitfields, or if there are no more free bits in the bitfield.

ppryga replied:
> But didn't we have that discussion in your earlier PR?

That is very possible. I've taken a look on the topic again and found better understanding.

Based on cppref (but in reference to C language ;-)): https://en.cppreference.com/w/c/language/bit_field

> Is that always 1 octet, or will it be 2 octets, and is it the same for all compilers?

`Declares a member with explicit width, in bits. Adjacent bit field members may be packed to share and straddle the individual bytes.` and `Multiple adjacent bit fields are permitted to be (and usually are) packed together`
This is related with all types.

What more: `Even though the number of bits in the object representation of _Bool is at least CHAR_BIT, the width of the bit field of type _Bool cannot be greater than1 (since C99).`

Summarizing, there is no guarantee that any bitfield will be packed together, even based on the same type. 
In case of gcc and clang in most of recent revisions merge bool with uint8_t fields.

```
struct some {
    bool a:1;
    uint16_t b:6;
    bool d:1;
    uint16_t c:1;
};
``` 
This struct has size of 2 when compiled with x86-gcc-11.2. Also this one has size of 2:

```
struct some {
    uint8_t a:1;
    uint16_t b:6;
    uint8_t d:1;
    uint16_t c:1;
};
```

Fortunately this is not a structure that isn't used to store PDUs data.

ppryga replied:
I've decided to change `bool` to `uint8_t` and assignments to these members to `1U/0U`. I'll stick to convention that boolean members in bitfields are also defined as `uint8_t` unless all members are booleans.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -1189,21 +1197,73 @@ uint8_t ll_df_set_conn_cte_req_enable(uint16_t handle, uint8_t enable, uint8_t c
+		return BT_HCI_ERR_UNKNOWN_CONN_ID;
+	}
+
+	if (!enable) {
+		/* There is no parameter validation for disable operation. */
+
+		/* TODO: Add missing implementation of disable CTE reques.
+		 * Requires refactored LLCPs.
+		 */
+	} else {
```
Suggest to switch around so that `if (enable)` comes first. At least to me it's makes more sense to do `if (true)` before `if (!true)` if we expect enable to be called as often with `0` as `1`.

ppryga replied:
Code reorganized.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_llcp.c:
```diff
@@ -452,6 +455,14 @@ void ull_llcp_init(struct ll_conn *conn)
 
+#if defined(CONFIG_BT_CTLR_DF_CONN_CTE_REQ)
+	conn->llcp.cte_req.is_enabled = false;
+	conn->llcp.cte_req.req_expire = 0;
```
```suggestion
	conn->llcp.cte_req.req_expire = 0U;
```

ppryga replied:
Corrected

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_llcp_common.c:
```diff
@@ -737,36 +750,40 @@ static void rp_comm_tx(struct ll_conn *conn, struct proc_ctx *ctx)
 		 * CTE is supported for PHY1M.
 		 */
-		if (conn->lll.phy_tx != PHY_CODED) {
+		if (conn->lll.phy_tx == PHY_CODED) {
```
Shouldn't this be 
```suggestion
		if (conn->lll.phy_tx != PHY_1M {
```
based on the comment?

ppryga replied:
The comment was wrong. It should say that the CTE is not supported with CODED PHY.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_llcp_pdu.c:
```diff
@@ -727,27 +727,46 @@ void llcp_pdu_encode_cte_req(struct proc_ctx *ctx, struct pdu_data *pdu)
 
-	/* TODO add handling of IQ samples forwarding */
+	/* Received LL_CTE_RSP PDU didn't have CTE */
+	pdu->cp = 0U;
```
I think earlier you used `true`/`false` for a `uint8_t:1`. How come you are using `0U` and `1U` for this one then?

ppryga replied:
We had a conversation about it already. The outcome was that I had used `bool` for `cp` bit but then it has been updated by @cvinayak to make it closer to specification. I'm using boolean in other places because they are not related with structures from BT spec. or hardware spec.

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -2949,7 +2949,40 @@ void bt_hci_le_df_connection_iq_report(struct net_buf *buf)
+		return;
+	}
+
+	for (cb = callback_list; cb; cb = cb->_next) {
+		if (cb->cte_report_cb) {
+			cb->cte_report_cb(conn, &iq_report);
+		}
+	}
+
+	STRUCT_SECTION_FOREACH(bt_conn_cb, cb)
+	{
+		if (cb->cte_report_cb) {
+			cb->cte_report_cb(conn, &iq_report);
+		}
+	}
```
Hmm, `cte_report_cb` doesn't have `iq_report` (or `info`) as `const`, so the receiver of the callbacks may actually change the data in the struct. Is that intentional?

ppryga replied:
It has `const`: https://github.com/zephyrproject-rtos/zephyr/blob/9d346cc2b248d40742f8633f8fc3f7842c44800a/include/bluetooth/bluetooth.h#L1256-L1257

`sync` pointer doesn't have `const`, as it hasn't got it in `state_changed`, `biginfo`...

Thalley replied:
Oh, my bad, I guess the placement of `const` made me miss it :D

Thalley commented on subsys/bluetooth/host/direction.c:
```diff
@@ -710,6 +711,46 @@ static int hci_df_set_conn_cte_req_enable(struct bt_conn *conn, bool enable,
+
+	evt = net_buf_pull_mem(buf, sizeof(*evt));
+
+	conn = bt_conn_lookup_handle(sys_le16_to_cpu(evt->conn_handle));
+
+	if (!conn) {
```
```suggestion
	conn = bt_conn_lookup_handle(sys_le16_to_cpu(evt->conn_handle));
	if (!conn) {
```

ppryga replied:
Corrected

Thalley replied:
Doesn't looks like you saved :)

ppryga replied:
Yes, sorry!

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -1189,21 +1197,73 @@ uint8_t ll_df_set_conn_cte_req_enable(uint16_t handle, uint8_t enable, uint8_t c
+		}
+
+#if defined(CONFIG_BT_CTLR_PHY)
+		/* Phy may be changed to CODED only if PHY update procedure is supproted. In other
```
Typo
```suggestion
		/* Phy may be changed to CODED only if PHY update procedure is supported. In other
```

ppryga replied:
The comment was wrong actually. I've changed it.

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -1189,21 +1197,73 @@ uint8_t ll_df_set_conn_cte_req_enable(uint16_t handle, uint8_t enable, uint8_t c
+	if (!enable) {
+		/* There is no parameter validation for disable operation. */
+
+		/* TODO: Add missing implementation of disable CTE reques.
```
typo
```suggestion
		/* TODO: Add missing implementation of disable CTE request.
```

ppryga replied:
Corected.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df.c:
```diff
@@ -307,17 +290,17 @@ static int init_reset(void)
  * @param cte_type   Type of the CTE
  * @param cte_length Length of CTE in units of 8 us
- * @param ant_ids_len Length of antenna identifiers array
+ * @param ant_sw_len Length of antenna switch pattern array
```
Yeah, it would be better to use same convention as in `direction.h` -> `num_ant_ids`. I have to align it in lll_df and ull_df but I'll do that in separate PR.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df_internal.h:
```diff
@@ -59,8 +59,9 @@ static inline uint8_t lll_df_sync_cfg_is_modified(struct lll_df_sync *df_cfg)
-/* Enable CTE transmission according to provided configuration */
-void lll_df_conn_cte_tx_enable(const struct lll_df_conn_tx_cfg *df_cfg);
+/* Configure CTE transmission */
+void lll_df_cte_tx_configure(uint8_t cte_type, uint8_t cte_length, uint8_t ant_ids_len,
```
Changed to `num_ant_ids`

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -1649,6 +1649,18 @@ void ull_conn_done(struct node_rx_event_done *done)
 #endif /* CONFIG_BT_CTLR_LE_PING */
 
+#if defined(CONFIG_BT_CTLR_DF_CONN_CTE_REQ)
+	if (conn->llcp.cte_req.req_expire != 0) {
+		if (conn->llcp.cte_req.req_expire > elapsed_event) {
+			conn->llcp.cte_req.req_expire -= elapsed_event;
+		} else {
+			conn->llcp.cte_req.req_expire = 0;
```
Corrected

ppryga-nordic committed:
Bluetooth: Controller: df: Add disable support for CTE REQ and RSP proc

Add missing implementation for disabling CTE request and resposne
control procedures.
If any of these commands is active in LLL then ULL context has
to wait before return. The wait mechanism is based on semaphore.
The semaphore is initialized in ULL context and given by code
responsible for command completion.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
@Thalley @kruithofa comments were addressed.
I've added another commit with  CTE REQ/RSP disable functionality.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.h:
```diff
@@ -11,16 +11,16 @@
-#define RADIO_PKT_CONF_PHY_1M (1U)
-#define RADIO_PKT_CONF_PHY_2M (2U)
-#define RADIO_PKT_CONF_PHY_CODED (3U)
+#define RADIO_PKT_CONF_PHY_1M (BIT(1U))
+#define RADIO_PKT_CONF_PHY_2M (BIT(2U))
+#define RADIO_PKT_CONF_PHY_CODED (BIT(3U))
```
Any reason to change the values?
```suggestion
#define RADIO_PKT_CONF_PHY_1M (BIT(0U))
#define RADIO_PKT_CONF_PHY_2M (BIT(1U))
#define RADIO_PKT_CONF_PHY_CODED (BIT(2U))
```

ppryga replied:
Thanks for spotting that wrong BIT macro values were used. The values were plain int values `1,2,3` that were wrong because PHYs are defined and used by controller as 
```
#define PHY_1M       BIT(0)
#define PHY_2M       BIT(1)
#define PHY_CODED    BIT(2)
```
hence the flags macros should reflect that.

ppryga replied:
Macros values corrected.

ppryga commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -7441,10 +7494,20 @@ static void encode_data_ctrl(struct node_rx_pdu *node_rx,
 		le_unknown_rsp(pdu_data, handle, buf);
 		break;
 
+	case PDU_DATA_LLCTRL_TYPE_REJECT_EXT_IND:
+		le_reject_ext_ind(pdu_data, handle, buf);
+		break;
+
```
Well, I have no strong opinion about that. The `le_unknown_rsp` does not assert in similar situation. In case there is a not handled rejection then there will be warning logged and nothing more happens. @thoh-ot @kruithofa what is your opinion about it?

cvinayak replied:
No strong change request here from me.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_llcp.c:
```diff
@@ -452,6 +455,20 @@ void ull_llcp_init(struct ll_conn *conn)
+#if defined(CONFIG_BT_CTLR_DF_CONN_CTE_RSP)
+	conn->llcp.cte_rsp.is_enabled = false;
+	conn->llcp.cte_rsp.is_active = false;
+	conn->llcp.cte_rsp.disable_param = 0U;
```
Corrected

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_llcp.c:
```diff
@@ -452,6 +455,20 @@ void ull_llcp_init(struct ll_conn *conn)
+	conn->llcp.cte_req.is_enabled = false;
+	conn->llcp.cte_req.req_expire = 0U;
+	conn->llcp.cte_req.is_active = false;
+	conn->llcp.cte_req.disable_param = 0U;
```
Corrected

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_conn_types.h:
```diff
@@ -422,19 +425,15 @@ struct llcp_struct {
+		uint16_t req_expire;
 	} cte_req;
+
 #endif /* CONFIG_BT_CTLR_DF_CONN_CTE_REQ */
```
Corrected

ppryga commented:
I've written to community on discord: https://discord.com/channels/720317445772017664/733037944922964069/933669431572328528 for help with failing tests.
It is strange that compiler can't find `cp` in `struct pdu_data` but it is able to find `cte_info` at the same time. The struct is defined in single place `pdu.h`. The tests are working on local machine.

ppryga commented:
@cvinayak your comments were resolved.

cvinayak commented:
> I've written to community on discord: https://discord.com/channels/720317445772017664/733037944922964069/933669431572328528 for help with failing tests. It is strange that compiler can't find `cp` in `struct pdu_data` but it is able to find `cte_info` at the same time. The struct is defined in single place `pdu.h`. The tests are working on local machine.

Did you try rebase the branch and build?

cvinayak reviewed (approved):
No strong changes requests from my end.

ppryga commented:
> Did you try rebase the branch and build?

I've just rebased it, let's see if it builds.

ppryga-nordic committed:
Bluetooth: Controller: Enable PDU CP bit when CTE RX is enabled

The CP bit is read in DF connected mode while interpreting
LL_CTE_RSP PDU, hence it must be available in struct pdu_data
type.

There were missing two Kconfig options in new LLCP tests.
They were reponsbile for disable of CP bit instruct pdu_data.
That caused tests to fail during compilation.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df.c:
```diff
@@ -307,17 +290,17 @@ static int init_reset(void)
- * @param ant_ids    Pointer to antenna identifiers array
+ * @param cte_type    Type of the CTE
+ * @param cte_length  Length of CTE in units of 8 us
+ * @param num_ant_ids Length of antenna switch pattern array
```
I went for second option.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df_types.h:
```diff
@@ -154,11 +149,13 @@ struct cte_conn_iq_report {
-	uint8_t cte_type:2;
-	uint8_t cte_length:6; /* Length of CTE in 8us units */
-	uint8_t cte_rsp_en:1; /* CTE response is enabled */
+	bool cte_rsp_en:1; /* CTE response is enabled */
```
Changed to `uint8_t`

ppryga commented:
@Thalley comments addressed.

I've added new commit with fix for failing tests.

ppryga commented:
I've found a bug in commit: 5b0f8af. Wrong CTE duration was set that caused dropping of a connection.

ppryga commented:
@nashif or @cfriedt could you merge this PR? It's blocking these two: 
https://github.com/zephyrproject-rtos/zephyr/pull/42003
https://github.com/zephyrproject-rtos/zephyr/pull/42003

ppryga commented:
Thanks!
